raft/src/main/java/org/apache/kafka/raft/internals/ReplicaKey.java

```diff
@@ -21,7 +21,9 @@
 import java.util.Objects;
 import java.util.Optional;
 
-public final class ReplicaKey {
+public final class ReplicaKey implements Comparable<ReplicaKey> {
+    public static final Uuid NO_DIRECTORY_ID = Uuid.ZERO_UUID;
+
     private final int id;
     private final Optional<Uuid> directoryId;
```

Comment:
We should use `Uuid.ZERO_UUID` in situations where there is no defined UUID, to avoid having so many Optionals. (This is how it's handled at the serde level anyway)

Reply:
Yes. We turn `Optional.empty()` to `Uuid.ZERO_UUID` when serializing.

I mainly set this as an `Optional<Uuid>` because KRaft needs to do some special handling when comparing `ReplicaKey` (id, directory id) and the voter set's directory id is set to `Optional.empty()/Uuid.ZERO_UUID`. I wanted to make it as obvious as possible that those cases need to be handle different and KRaft can't just compare `Uuid`s.

I file an issue to revisit this. [Consider replacing the directory id Optional<Uuid> with just Uuid](https://issues.apache.org/jira/browse/KAFKA-17033). Maybe the better solution is to introduce a simple type like `class DirectoryId(Uuid id)` that encapsulates this special handling.

Reply:
I don't think having a special class would be better (in fact, I think it would be strictly worse than just using `Optional`)

If you feel strongly about this, then I guess just leave it as Optional. I find this messy (for one thing, you have to make sure to remember to transform UUID_ZERO into Optional.empty in all the relevant places) but maybe I'm not going to convince you here :)